### ISSUE TITLE
Different loss value based on shape

### ISSUE THREAD

**Issue by devansh20la** [pytorch/pytorch#19699]
## 🐛 Bug
MSE Loss between tensor A with shape (N, 1), and tensor B with shape (N) 
is different than loss between tensor A with shape (N,1) and tensor B with shape(N,1).

## To Reproduce

```
import torch
loss = torch.nn.MSELoss()
a = torch.randn(100,1)
b = torch.randn(100)
loss(a,b)
>>> tensor(2.0373)
loss(a,b.reshape(-1,1))
>>>  tensor(1.7123)
```

## Expected behavior
Same results or at least an error 

## Environment
```
Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 9.1.85
GPU models and configuration: 
GPU 0: GeForce GTX 1080 Ti
GPU 1: GeForce GTX 1080 Ti
GPU 2: GeForce GTX 1080 Ti
GPU 3: GeForce GTX 1080 Ti

Nvidia driver version: 390.30
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy==1.16.2
[pip] numpydoc==0.8.0
[pip] torch==1.0.1.post2
[pip] torchfile==0.1.0
[pip] torchvision==0.2.2
[conda] blas                      1.0                         mkl  
[conda] mkl                       2019.3                      199  
[conda] mkl-service               1.1.2            py37h90e4bf4_5  
[conda] mkl_fft                   1.0.10           py37ha843d7b_0  
[conda] mkl_random                1.0.2            py37hd81dba3_0  
[conda] pytorch                   1.0.1           py3.7_cuda9.0.176_cudnn7.4.2_2    pytorch
[conda] torchfile                 0.1.0                    pypi_0    pypi
[conda] torchvision               0.2.2                      py_3    pytorch
```

**Comment by soumith** [comment]
this is following standard broadcasting semantics, and so the Tensors in the first case broadcast to `N, N)`

We recognized that this is highly unintuitive and now issue a warning when two Tensors that broadcast in loss functions.
This warning will be part of the next release and was fixed via https://github.com/pytorch/pytorch/pull/18349
